wordpress-mobile/WordPress-iOS
# Fix a crash due to Core Data Concurrency Violation in the Weekly Roundup Background Task

Fixes #20957

## Description

This PR rectifies a crash that occurs during the Weekly Roundup Background Task, resulting from a Core Data Concurrency Violation.

### Root of the Problem

The principal cause of this crash stems from the Weekly Roundup task's method of fetching a list of blogs. It uses the main context, however, this is done from a background thread, resulting in a direct Core Data Concurrency Violation. A screenshot illustrating the issue is included below.

![Issue Screenshot](https://github.com/wordpress-mobile/WordPress-iOS/assets/9609223/5d8ae737-bb3b-4de0-992f-21ae05e7c6bd)

### Proposed Solution

- Modify the code segment that generates the list of blogs so that it conducts the fetch on a derived context's queue.
- Ensure that users of the blogs list access the blogs from the context's queue.
- For those users who no longer have access to the context that initiated the blogs, they will be passed the blogs `NSManagedObjectID`, which will facilitate the creation of new blog instances.

## Test Instructions

<details>
  <summary>How to trigger Weekly Roundup Task</summary>

1. Head to Me > App Settings > Debug > Weekly Roundup then tap Schedule Immediately
2. If the above doesn't work for you, and it didn't work for me even on `trunk`, you can run this code on `AppDelegate`.

```swift
let task = WeeklyRoundupBackgroundTask()
task.run { error in
    DDLogError(error)
} completion: { completed in
    DDLogInfo("Weekly Roundup Background Task Completed: \(completed)")
}
```
</details>

I recommend to perform the following tests on `trunk` to verify the before and after behavior. 

#### Testing whether the crash is resolved

1. Make sure `-com.apple.CoreData.ConcurrencyDebug` is enabled.
2. Build and run Jetpack app.
3. Trigger the the Weekly Round up Task. See **How to trigger Weekly Roundup Task** above.
4. The app shouldn't crash.

#### Testing whether the push notifications appear

1. Make sure you have access to Weekly Roundup enabled sites. 
2. Trigger the the Weekly Round up Task.
3. Wait for a few seconds.
4. **Expect** to receive up to 5 push notifications
5. Tap one of the push notifications.
6. **Expect** to be redirected to Stats screen.

## Regression Notes
1. Potential unintended areas of impact
The Weekly Roundup feature should be re-tested.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Existing Weekly Roundup unit tests and manual testing.

3. What automated tests I added (or what prevented me from doing so)
None. 

## PR submission checklist:
